File: prioritarr/clients/sonarr.py
from __future__ import annotations

import httpx


class SonarrClient:
    """Thin synchronous wrapper around the Sonarr v3 REST API."""

    def __init__(self, base_url: str, api_key: str, timeout: float = 120.0) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = httpx.Client(
            headers={"X-Api-Key": api_key},
            timeout=timeout,
        )

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _get(self, path: str, params: dict | None = None) -> object:
        response = self._client.get(f"{self._base_url}{path}", params=params)
        response.raise_for_status()
        return response.json()

    def _post(self, path: str, json_data: dict) -> object:
        response = self._client.post(f"{self._base_url}{path}", json=json_data)
        response.raise_for_status()
        return response.json()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get_all_series(self) -> list:
        return self._get("/api/v3/series")

    def get_series(self, series_id: int) -> dict:
        return self._get(f"/api/v3/series/{series_id}")

    def get_episodes(self, series_id: int) -> list:
        return self._get("/api/v3/episode", params={"seriesId": series_id})

    def get_wanted_missing(self, page_size: int = 1000) -> list:
        data = self._get("/api/v3/wanted/missing", params={"pageSize": page_size})
        return data["records"]

    def get_wanted_cutoff(self, page_size: int = 1000) -> list:
        data = self._get("/api/v3/wanted/cutoff", params={"pageSize": page_size})
        return data["records"]

    def get_queue(self, page_size: int = 1000) -> list:
        data = self._get("/api/v3/queue", params={"pageSize": page_size})
        return data["records"]

    def trigger_series_search(self, series_id: int) -> dict:
        return self._post(
            "/api/v3/command",
            {"name": "SeriesSearch", "seriesId": series_id},
        )

    def trigger_cutoff_search(self, series_id: int) -> dict:
        return self._post(
            "/api/v3/command",
            {"name": "CutoffUnmetEpisodeSearch", "seriesId": series_id},
        )

File: prioritarr-kotlin/app/src/main/kotlin/org/cquemin/prioritarr/clients/Sonarr.kt
package org.cquemin.prioritarr.clients

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.put

/** Thin wrapper around the Sonarr v3 REST API. Mirrors prioritarr/clients/sonarr.py. */
class SonarrClient(
    private val baseUrl: String,
    private val apiKey: String,
    private val http: HttpClient,
) {
    private val root: String = baseUrl.trimEnd('/')

    suspend fun getAllSeries(): JsonArray =
        get("/api/v3/series").jsonArray

    suspend fun getSeries(seriesId: Long): JsonObject =
        get("/api/v3/series/$seriesId") as JsonObject

    suspend fun getEpisodes(seriesId: Long): JsonArray =
        get("/api/v3/episode", mapOf("seriesId" to seriesId.toString())).jsonArray

    suspend fun getWantedMissing(pageSize: Int = 1000): JsonArray =
        ((get("/api/v3/wanted/missing", mapOf("pageSize" to pageSize.toString())) as JsonObject)
            ["records"] as JsonArray)

    suspend fun getWantedCutoff(pageSize: Int = 1000): JsonArray =
        ((get("/api/v3/wanted/cutoff", mapOf("pageSize" to pageSize.toString())) as JsonObject)
            ["records"] as JsonArray)

    suspend fun getQueue(pageSize: Int = 1000): JsonArray =
        ((get("/api/v3/queue", mapOf("pageSize" to pageSize.toString())) as JsonObject)
            ["records"] as JsonArray)

    suspend fun triggerSeriesSearch(seriesId: Long): JsonObject =
        post("/api/v3/command", buildJsonObject {
            put("name", "SeriesSearch")
            put("seriesId", seriesId)
        }) as JsonObject

    suspend fun triggerCutoffSearch(seriesId: Long): JsonObject =
        post("/api/v3/command", buildJsonObject {
            put("name", "CutoffUnmetEpisodeSearch")
            put("seriesId", seriesId)
        }) as JsonObject

    private suspend fun get(path: String, params: Map<String, String> = emptyMap()): JsonElement =
        http.get("$root$path") {
            header("X-Api-Key", apiKey)
            for ((k, v) in params) parameter(k, v)
        }.body()

    private suspend fun post(path: String, body: JsonElement): JsonElement =
        http.post("$root$path") {
            header("X-Api-Key", apiKey)
            contentType(ContentType.Application.Json)
            setBody(body)
        }.body()
}
